### PR TITLE
Laboratory work #3, Tatiana Zheleznova - 22FPL2

### PR DESCRIPTION
**Pull request by tanya-iron**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -270,6 +369,22 @@ def run(self, seq_len: int, prompt: str) -> Optional[str]:
...
+            return None
+        encoded = self._text_processor.encode(prompt)
+        n_gram_size = self._model.get_n_gram_size()
+        if not (encoded and n_gram_size):
+            return None
+        for i in range(seq_len):
+            next_tokens = self._model.generate_next_token(encoded[-(n_gram_size + 1):])
+            if not next_tokens:
+                break
+            best_next = [token for token, freq in next_tokens.items() if freq == max(next_tokens.values())]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:382]: sorted by freq and get first - it is more clear

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -41,6 +43,18 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str):
+            return None
+        tokenized_text = []
+        for token in text.lower():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:49]: too dificult

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py]: I changed it, so fixed?

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py:50]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -76,6 +97,11 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+        for key, value in self._storage.items():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:102]: filter

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -93,6 +119,19 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
...
         """
+        if not isinstance(text, str) or len(text) == 0:
+            return None
+        tokenized_text = self._tokenize(text)
+        if not tokenized_text:
+            return None
+        for token in tokenized_text:
+            self._put(token)
+        encoded_text = []
+        for token in tokenized_text:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:130]: the same for. merge with loop above

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -104,6 +143,10 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+        if element not in self._storage:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:148]: early return. + looks like pylint error

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -143,6 +195,14 @@ def _decode(self, corpus: tuple[int, ...]) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(corpus, tuple) or len(corpus) == 0:
+            return None
+        decoded_text = []
+        for token in corpus:
+            if not self.get_token(token):
+                return None
+            decoded_text.append(self.get_token(token))
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:204]: call `self.get_token(token)` twice. Why?

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -159,6 +219,16 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+        if not isinstance(decoded_corpus, tuple) or len(decoded_corpus) == 0:
+            return None
+        list_tokens = []
+        for i in decoded_corpus:
+            list_tokens += i
+        if list_tokens[-1] == self._end_of_word_token:
+            del list_tokens[-1]
+        new_line = ''.join(list_tokens)
+        new_line = new_line.replace(self._end_of_word_token, ' ')
+        return new_line.capitalize() + '.'
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:231]: fstring

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -208,6 +282,25 @@ def build(self) -> int:
...
+        if not (isinstance(self._encoded_corpus, tuple) and self._encoded_corpus):
+            return 1
+        n_grams = self._extract_n_grams(self._encoded_corpus)
+        if not (isinstance(n_grams, tuple) and n_grams):
+            return 1
+        contexts = {}
+        for n_gram in n_grams:
+            if not isinstance(n_gram, tuple):
+                return 1
+            if n_gram[:-1] not in contexts:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:294]: 1

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -208,6 +282,25 @@ def build(self) -> int:
...
+                contexts[n_gram[:-1]] = {}
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:295]: 2

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -208,6 +282,25 @@ def build(self) -> int:
...
+            if n_gram not in contexts[n_gram[:-1]]:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:296]: 3

File: lab_3_generate_by_ngrams/main.py [hunk 11]
```diff
@@ -208,6 +282,25 @@ def build(self) -> int:
...
+                contexts[n_gram[:-1]][n_gram] = 0.0
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:297]: 4

File: lab_3_generate_by_ngrams/main.py [hunk 12]
```diff
@@ -208,6 +282,25 @@ def build(self) -> int:
...
+            contexts[n_gram[:-1]][n_gram] += 1.
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:298]: 5

Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:298]: n_gram[:-1] - five times!

File: lab_3_generate_by_ngrams/main.py [hunk 13]
```diff
@@ -270,6 +377,28 @@ def run(self, seq_len: int, prompt: str) -> Optional[str]:
...
+        n_gram_size = self._model.get_n_gram_size()
+        encoded = self._text_processor.encode(prompt)
+        if not (encoded and n_gram_size):
+            return None
+        for iteration in range(seq_len):
+            tokens = self._model.generate_next_token(encoded[-n_gram_size + 1:])
+            if not tokens:
+                break
+            max_freq = max(tokens.values())
+            biggest_token = [candidate for candidate, freq in tokens.items()
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:391]: filter

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py:204]: To check if it doesn't return None? What is the alternative?

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 4 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py:128]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py]: fixed?

File: lab_3_generate_by_ngrams/main.py (hunk 7 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 12 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py]: fixed

Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py]: ?

File: lab_3_generate_by_ngrams/main.py (hunk 13 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py]: fixed?

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:199]: save to a variable?

File: lab_3_generate_by_ngrams/main.py [hunk 14]
```diff
@@ -159,6 +214,14 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or len(decoded_corpus) == 0:
+            return None
+        list_tokens = []
+        for i in decoded_corpus:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:220]: bad name  for i

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 14 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (tanya-iron) [reply lab_3_generate_by_ngrams/main.py]: fixed

**Pull request closed without merging**